client/downloader.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2021-2022, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// filePartDescriptor defines one part of multipart download.
+type filePartDescriptor struct {
+	start int64
+	end   int64
+	cur   int64
+
+	w io.WriterAt
+}
+
+// Write writes buffer 'p' at offset 'start' using 'WriteAt()' to atomically seek and write.
+// Returns bytes written
+func (ps *filePartDescriptor) Write(p []byte) (n int, err error) {
+	n, err = ps.w.WriteAt(p, ps.start+ps.cur)
+	ps.cur += int64(n)
+
+	return
+}
+
+// minInt64 returns minimum value of two arguments
+func minInt64(a, b int64) int64 {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+const (
+	defaultStreams  uint = 4
+	defaultPartSize      = 5 * 1024 * 1024
+)
+
+// Download performs download of contents at url by writing 'size' bytes to 'dst' using credentials 'c'.
+func (c *Client) multipartDownload(ctx context.Context, u string, creds credentials, w io.WriterAt, size int64, spec *Downloader, pb ProgressBar) error {
+	if size <= 0 {
+		return fmt.Errorf("invalid image size (%v)", size)
+	}
+
+	// Initialize the progress bar using passed size
+	pb.Init(size)
+
+	// Clean up (remove) progress bar after download
+	defer pb.Wait()
+
+	// Calculate # of parts
+	parts := uint(1 + (size-1)/spec.PartSize)
+
+	c.Logger.Logf("size: %d, parts: %d, streams: %d, partsize: %d", size, parts, spec.Concurrency, spec.PartSize)
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	// Allocate channel for file part requests
+	ch := make(chan filePartDescriptor, parts)
+
+	// Create download part workers
+	for n := uint(0); n < defaultStreams; n++ {
```

Comment:
Does this mean we are always using 4 streams, despite the fact we are logging a spec.Concurrency above?

Reply:
Yes it does... I'll fix it :)